### PR TITLE
SD-712 Add eXclusiv OR conditions between CQR, SCR.

### PR DESCRIPTION
**Pull request by preetamnpr**

### CODE REVIEW

File: booking/src/main/java/org/dcsa/conformance/standards/booking/checks/BookingChecks.java [hunk 1]
```diff
@@ -524,11 +524,19 @@ private static void generateScenarioRelatedChecks(List<JsonContentCheck> checks,
...
-      "contractQuotationReference",
-      delayedValue(cspSupplier, CarrierScenarioParameters::contractQuotationReference)
-    ));
+      body -> {
+        var contractQuotationReference = body.path("contractQuotationReference").asText("");
+        var serviceContractReference = body.path("serviceContractReference").asText("");
+        if (!contractQuotationReference.isEmpty() && !serviceContractReference.isEmpty()) {
+          return Set.of("The scenario requires either of 'contractQuotationReference'/'serviceContractReference'" +
+            " to be present, but not both");
+        }
```
[the remark this answers is not in the document]
Comment (preetamnpr) [reply booking/src/main/java/org/dcsa/conformance/standards/booking/checks/BookingChecks.java:536]: if neither of them are present then Set.of() is returned, as these fields are not mandatory. This check is used to check the Scenario input values are not been added in each use case actions, as input is fed by the user.

File: booking/src/main/java/org/dcsa/conformance/standards/booking/checks/BookingChecks.java [hunk 2]
```diff
@@ -524,11 +524,19 @@ private static void generateScenarioRelatedChecks(List<JsonContentCheck> checks,
...
       delayedValue(cspSupplier, CarrierScenarioParameters::carrierServiceName)
     ));
-    checks.add(JsonAttribute.mustEqual(
+
+    checks.add(JsonAttribute.customValidator(
       "[Scenario] Verify that the correct 'contractQuotationReference' is used",
-      "contractQuotationReference",
-      delayedValue(cspSupplier, CarrierScenarioParameters::contractQuotationReference)
-    ));
+      body -> {
```
Comment (palatsangeetha) [inline booking/src/main/java/org/dcsa/conformance/standards/booking/checks/BookingChecks.java:530]: Should we have in the Scenario message also correct 'contractQuotationReference'/'serviceContractReference' is used?

File: core/pom.xml [hunk 3]
```diff
@@ -28,11 +28,29 @@
...
+			<artifactId>mockito-junit-jupiter</artifactId>
+			<version>5.6.0</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.mockito</groupId>
+			<artifactId>mockito-junit-jupiter</artifactId>
+			<version>5.11.0</version>
+			<scope>test</scope>
+		</dependency>
```
Comment (jkosternl) [inline core/pom.xml:54]: Likely we only need 1 dependency here, which is `mockito-core` and please use the latest.

File: core/src/test/java/org/dcsa/conformance/core/check/JsonAttributeTest.java [hunk 4]
```diff
@@ -0,0 +1,732 @@
...
+  }
+
+  @Test
+  void testContentChecksWithList() {
+    Predicate<String> predicate = s -> true;
+    assertDoesNotThrow(
+        () ->
+            JsonAttribute.contentChecks(
+                predicate, UUID.randomUUID(), HttpMessageType.REQUEST, "2.0.0", jsonContentCheck));
+  }
```
Comment (jkosternl) [inline core/src/test/java/org/dcsa/conformance/core/check/JsonAttributeTest.java:66]: This unit test is the same as the above. Please adjust or remove.

File: core/src/test/java/org/dcsa/conformance/core/check/JsonAttributeTest.java [hunk 5]
```diff
@@ -0,0 +1,732 @@
...
+  @ParameterizedTest(name = "[{index}] {0}")
+  @MethodSource("testIsNotNullArgs")
+  void testIsNotNull(String testName, ObjectNode input, String path, boolean expected) {
+    assertEquals(expected, JsonAttribute.isNotNull(JsonPointer.compile(path)).test(input));
+  }
+
+  @Test
+  void testMustBeNotNull() {
+    ObjectNode objectNode = JsonNodeFactory.instance.objectNode();
+    objectNode.put("test", "test");
```
Comment (jkosternl) [inline core/src/test/java/org/dcsa/conformance/core/check/JsonAttributeTest.java:174]: These 2 sentences are repeated 18 times. Please put them in a `before()` method, so JUnit runs them automatically before every invocation of a test. And make `objectNode` a variable of this class.

File: core/src/test/java/org/dcsa/conformance/core/check/JsonAttributeTest.java [hunk 6]
```diff
@@ -0,0 +1,732 @@
...
+  }
+
+  @Test
+  void testLostAttributeCheck() {
+    ObjectNode objectNode = JsonNodeFactory.instance.objectNode();
+    objectNode.put("test", "test");
+    Supplier<JsonNode> supplier = () -> objectNode;
+
+    assertDoesNotThrow(
+        () -> org.dcsa.conformance.core.check.JsonAttribute.lostAttributeCheck("test", supplier));
```
Comment (jkosternl) [inline core/src/test/java/org/dcsa/conformance/core/check/JsonAttributeTest.java:235]: Is the full package reference needed? Same for line 240.
```suggestion
        () -> JsonAttribute.lostAttributeCheck("test", supplier));
```

**Review by jkosternl — APPROVED** [review]
Looking good Preetam. Nice that you covered all methods. Added a few minor remarks (as a few we've already talked about).

File: core/src/test/java/org/dcsa/conformance/core/check/JsonAttributeTest.java (hunk 5 above)
Comment (preetamnpr) [reply core/src/test/java/org/dcsa/conformance/core/check/JsonAttributeTest.java:174]: ObjectNode objectNode = JsonNodeFactory.instance.objectNode(); line is moved to BeforeEach method.

**Pull request merged**